### PR TITLE
[Quant] Add experimental online LUT-B quantization for MoE

### PR DESCRIPTION
**Pull request by mgoin**
## Summary

- add `lut_b_moe` online quantization for routed MoE expert weights only
- fit one eight-entry E4M3 codebook per 8x64 tile and store 3-bit indices
- decode inside the Triton MoE GEMM for small batches; at `M >= 2 * local_experts`, dequantize with Triton and use the unquantized grouped GEMM
- leave dense linear layers unquantized and omit residual/scaling variants

This supersedes #50168 with a narrower approach. That draft targets dense
linear layers and explores residual/scaling algorithms; this PR targets only
routed MoE weights and includes executable MoE kernels.

## Validation

- `CUDA_VISIBLE_DEVICES=0 .venv/bin/python -m pytest tests/quantization/test_online.py::test_lut_b_stacked_moe_weight_format_and_fit tests/quantization/test_online.py::test_lut_b_preserves_zero_padding tests/kernels/moe/test_lut_b_moe.py -q -s` — 8 passed
- `CUDA_VISIBLE_DEVICES=0 .venv/bin/python -m pytest 'tests/quantization/test_online.py::test_online_quantization[False-lut_b_moe-None-UnquantizedLinearMethod-LutBOnlineMoEMethod]' -q -s` — passed
- full pre-commit hooks on changed files — passed
- `Qwen/Qwen3.6-35B-A3B`, TP=2: 256-request smoke passed with 512 prompt tokens and 256 output tokens
- `Qwen/Qwen3.6-35B-A3B`, TP=2: the first 100 GSM8K 5-shot questions scored 0.84 for LUT-B versus 0.83 for the BF16 Triton reference, with zero invalid answers for both

This PR was developed with AI assistance. The human submitter will review
every changed line before marking it ready for review.

## Summary by CodeRabbit

* **New Features**
  * Added `lut_b_moe` quantization support for routed MoE expert weights.
  * Added online quantization with 3-bit indices and E4M3 codebooks.
  * Added fused execution and automatic dequantization paths for improved MoE processing.
  * Added CLI configuration and a Quick Start example for serving compatible Qwen models.

* **Documentation**
  * Documented the supported scheme, format, activation types, and execution behavior.

### CODE REVIEW

File: vllm/model_executor/layers/quantization/online/lut_b.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+class LutBOnlineMoEMethod(OnlineMoEMethodBase):
+    """Quantize BF16/FP16 routed experts to LUT-B during model loading."""
+
+    def __init__(self, *, layer: torch.nn.Module):
+        if not current_platform.is_cuda_alike():
+            raise ValueError("LUT-B online MoE quantization requires a GPU.")
+        if layer.moe_config.has_bias:
+            raise ValueError("LUT-B online MoE does not currently support bias.")
+        super().__init__(layer.moe_config)
```
Comment (yewentao256) [inline vllm/model_executor/layers/quantization/online/lut_b.py:36]: Does this work in previous architecture like SM80? Should check if not

File: vllm/model_executor/layers/quantization/online/lut_b.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        if getattr(layer, "_already_called_process_weights_after_loading", False):
+            return
+
+        self._zero_padding(layer)
+        w13, w13_codebook = quantize_lut_b(layer.w13_weight)
+        w2, w2_codebook = quantize_lut_b(layer.w2_weight)
+        replace_parameter(layer, "w13_weight", w13.contiguous())
+        replace_parameter(layer, "w13_weight_codebook", w13_codebook.contiguous())
+        replace_parameter(layer, "w2_weight", w2.contiguous())
+        replace_parameter(layer, "w2_weight_codebook", w2_codebook.contiguous())
```
Comment (yewentao256) [inline vllm/model_executor/layers/quantization/online/lut_b.py:97]: Please take a look here, will zero padding after quantization still be 0 always?

File: vllm/model_executor/layers/fused_moe/experts/lut_b_moe.py [hunk 3]
```diff
@@ -0,0 +1,513 @@
...
+        quant_config=quant_config,
+        routing_tables=routing_tables,
+        allow_new_interface=True,
+        use_monolithic=False,
+    )
+    assert prepare_finalize is not None
+    experts = LutBTritonExperts(
+        moe_config=moe_config,
+        quant_config=quant_config,
+    )
```
Comment (yewentao256) [inline vllm/model_executor/layers/fused_moe/experts/lut_b_moe.py:549]: check is_supported_config?

**Review by yewentao256 — COMMENTED** [review]
Thanks for the work!

File: vllm/model_executor/layers/quantization/online/lut_b.py (hunk 1 above)
Comment (mgoin) [reply vllm/model_executor/layers/quantization/online/lut_b.py:36]: Yes it just uses a triton backend for fused weight-only dequant now since we don't have rubin kernels support yet

File: vllm/model_executor/layers/quantization/online/lut_b.py (hunk 2 above)
Comment (mgoin) [reply vllm/model_executor/layers/quantization/online/lut_b.py:97]: Good catch. `_zero_padding` ran before unconstrained fitting, so zeros in a mixed boundary tile could reconstruct as nonzero. I fixed this by pinning the closest codebook center to exact zero during every Lloyd iteration for tiles containing zeros; tiles without zeros still fit all eight entries freely. Added an adversarial mixed-tile regression test.

File: vllm/model_executor/layers/fused_moe/experts/lut_b_moe.py (hunk 3 above)
Comment (mgoin) [reply vllm/model_executor/layers/fused_moe/experts/lut_b_moe.py:549]: Added the check before construction. `LutBTritonExperts.is_supported_config` now carries the LUT-B bias and dimension constraints on top of the inherited device/activation/parallel/quantization checks, and `make_lut_b_moe_kernel` fails with the reported reason when unsupported. Added a regression test for this factory boundary.

File: vllm/model_executor/layers/quantization/online/lut_b.py (hunk 1 above)
Comment (yewentao256) [reply vllm/model_executor/layers/quantization/online/lut_b.py:36]: I meant fp8e4nv? Maybe add a check that this should be applied to SM90+

File: vllm/model_executor/layers/fused_moe/experts/lut_b_moe.py [hunk 4]
```diff
@@ -0,0 +1,550 @@
...
+        maybe_make_prepare_finalize,
+    )
+
+    supported, reason = LutBTritonExperts.is_supported_config(
+        LutBTritonExperts,
+        moe_config,
+        kLutBStatic,
+        None,
+        mk.FusedMoEActivationFormat.Standard,
+    )
```
Comment (yewentao256) [inline vllm/model_executor/layers/fused_moe/experts/lut_b_moe.py:534]: Do we support DeepEP-LL/NIXL as they return BatchedExperts?

Comment (mgoin) [reply vllm/model_executor/layers/fused_moe/experts/lut_b_moe.py:534]: This is not intended to be a production or performant backend, this is just meant for accuracy evaluation through emulation to prepare for Rubin support. So it is out of scope.

**Comment by mgoin** [comment]
@coderabbitai full review
